### DISCUSSION TITLE
Allow self-closing `/>` on HTML via a flag or feature

### DISCUSSION THREAD

**Discussion by phoenix-ru** [swc-project/swc#8424] [RFC]
Syntax like that is often used in Vue templates:
```html
<template>
  <slot />
  <div class="bad" />
</template>
```

Since `swc_html_parser` is spec-compliant, it acknowledges the `<slot />` and `<div />`, but does not actually pop the DOM Tree stack.
It makes parsing such templates impossible: https://github.com/phoenix-ru/fervid/issues/6

Can we add a flag to `ParserConfig` or a crate feature to treat such starting tags as self-closing? I can open a ticket and work on a PR.

**Comment by kdy1** [comment]
@alexander-akait Thoughts?

**Reply by alexander-akait** [reply]
@phoenix-ru @kdy1 hm, we can do it under the flag, but I am afraid vue syntax is more than `<tag />`, and so in future we will need more changes in our parser to support vue template syntax and it can be a problem... I am not familiar deeply with vue syntax, but I think basic SAX parser will be better here, for example https://github.com/fatihpense/rust_xml_oxide, HTML is not easy and keeping compatibility HTML and templates will increase complex of code very, but I am open to any feedback

**Reply by kdy1** [reply]
Is there a full list of custom syntaxes?

If you have one I may be able to design a trait-based extension API quickly

**Reply by alexander-akait** [reply]
@kdy1 

> Is there a full list of custom syntaxes?

I am afraid supporting all templates syntax is a complex task and I think we really don't need it

> If you have one I may be able to design a trait-based extension API quickly

In theory and ideally we can implement SAX based parser (like in link above), it can reuse some code from swc_html and swc_xml and provide authors to build own custom templates, but I think we need to have discuss here, do we really need to have it in swc and how many people will use it

**Reply by kdy1** [reply]
I'll provide only one trait and a generic arguemnt

**Reply by kdy1** [reply]
But well, I think we should consider `how many people will use it`. If the vue core team is going to use SWC's parsers, I think it worths the effort, but not otherwise

**Reply by alexander-akait** [reply]
Yeah, that is why, I think, we need more votes here

**Reply by phoenix-ru** [reply]
@alexander-akait @kdy1 There's no need to do anything beyond this little change.
Vue template syntax is complex enough, and I have already built a SWC-based parser for it.
To clarify — all the bindings (`:foo="bar"`), directives (`v-dir.mod1:arg="value"`) and event handlers (`@smth-changed`) are read correctly by `swc_html_parser` as raw html attributes. They are later parsed during AST transformation.
Supporting Vue in SWC is not needed, there's nothing out-of-spec going on except self-closed tags (which is legacy parser implementation really, but it's widely used).

**Reply by alexander-akait** [reply]
@phoenix-ru What we should do with `<br />` and `<br><br/>`?

**Reply by phoenix-ru** [reply]
@alexander-akait Afaik, br is a void tag. Nothing changes for it?

**Reply by alexander-akait** [reply]
@phoenix-ru I ask this, because I want to undestand what vue do with `<br>text</br>`, sorry I made typo

**Reply by phoenix-ru** [reply]
It would be great if, when flag is set, instead of reporting an error for `<div />` it would close it and continue. Same for other tags.

**Reply by phoenix-ru** [reply]
> @phoenix-ru I ask this, because I want to undestand what vue do with `<br>text</br>`, sorry I made typo

It is **invalid** syntax: https://play.vuejs.org/#eNp9kD1vwjAQhv+KdTNKVLEhK1JbMbRDW7UdvaThGgL+kn2GSJH/O7YRkAGx2e/znvWcJ3i2tjoEhBVwQmVlS9gIzRjfPjXTxJTvWYy8TreS/rmGcCRep4PQvJ7NwALId0b/D32180anJ6c8I6Azyg4S3aelwWgvYMUKyayV0hzfS0Yu4OKSd1vs9nfynR9zJuDLoUd3QAFXRq3rkc54/fORPGdQmU2Qqf0AfqM3MmTHc+0l6E3SnvWK7ZuyxtGg+1+/Hgm1vyyVRXMzlr6A9K+vD1a/6S6rZZkTOkI8Ad0Egzc=

**Reply by alexander-akait** [reply]
@phoenix-ru For any other tag? I just want to undestand what we should do with `<unknown />` and for `<br>` (without slash) and for `<br>text</br>`

**Reply by phoenix-ru** [reply]
Also invalid as per Nu HTML checker
![image](https://github.com/swc-project/swc/assets/18054980/5c007247-a07d-4ef9-9fad-0ed07786c922)

**Reply by phoenix-ru** [reply]
`<unknown>text</unknown>` is VALID (regular html element)
`<unknown />` is VALID (in Vue; proposed change to SWC)
`<br>` is VALID (it's a void tag)
`<br>text</br>` is INVALID (void tag, cannot close it)

**Reply by phoenix-ru** [reply]
I initially noticed that SWC has a `is_self_closing` attribute
https://rustdoc.swc.rs/swc_html_ast/struct.Element.html

Which is set to `true` for `<div />`.

**Reply by alexander-akait** [reply]
Okay, sounds good for me, do you want to send a PR? I will review to avoid regression in other places (but yeah, we have a lot of tests)

**Reply by phoenix-ru** [reply]
@alexander-akait Great then, I will try to solve it and open a PR.

**Comment by phoenix-ru** [comment]
Relevant place in Vue docs: https://vuejs.org/guide/essentials/component-basics.html#self-closing-tags.

I managed to get it working for "other" start tags (e.g. `<slot />`) by adding logic here:
https://github.com/swc-project/swc/blob/c39e75c61d4e2052e584993f0983ab515fbe5d61/crates/swc_html_parser/src/parser/mod.rs#L4115-L4118

Regarding normal HTML tags like `<p>`, `<b>`, `<div>` and others, I am not really sure how to proceed, since they all have additional spec baggage attached and simply popping the stack would not be correct.

Which is why I think it's better to do this, since it can be neatly organized as a macro and accurately added where necessary:
```rs
if self.config.allow_self_closing && *is_self_closing {
    let mut end_token_and_info = TokenAndInfo {
        span: token_and_info.span,
        acknowledged: false,
        token: Token::EndTag {
            tag_name: tag_name.to_owned(),
            raw_tag_name: None,
            is_self_closing: false,
            attributes: vec![],
        },
    };
    self.process_token(&mut end_token_and_info, None)?;
}
```

To make it possible, I changed the signatures of `token_and_info` to `&TokenAndInfo` in these two places, because there is no need for it to be mutable:
https://github.com/swc-project/swc/blob/c39e75c61d4e2052e584993f0983ab515fbe5d61/crates/swc_html_parser/src/parser/mod.rs#L8337-L8376

@alexander-akait Could you please comment on that?

**Reply by alexander-akait** [reply]
Let's try this solution, I don't remember code deeply, because it was a long time ago, but we will add many tests to ensure we have a right solution

**Reply by phoenix-ru** [reply]
I discovered a code smell in parser, which always returns `Result<T, Error>` regardless if `Err` variant is ever constructed.
The example is this function, which always returns `Ok` or uses `unreachable!`. This leads to `Result`s being used fanatically without any reason.

https://github.com/swc-project/swc/blob/c39e75c61d4e2052e584993f0983ab515fbe5d61/crates/swc_html_parser/src/parser/mod.rs#L8011-L8134

I am not touching it now, but I can create an issue (and perhaps tackle it later). Should I?
